llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -169,23 +169,15 @@ void ModuleHashGenerator::calculateGONumAndGOIMap(const Module &M) {
 #endif
 }
 
-void ModuleHashGenerator::updateDigestUseContributions(
-    MD5 &GOHash, const GOVec &Contributions) {
-  for (const GlobalObject *const G : Contributions) {
-    GOHash.update(GOIMap[G].InitialDigest.Bytes);
-  }
-  // GOHash/Module is changed if Contributions is not empty.
-  Changed = Changed || !Contributions.empty();
-}
-
-auto ModuleHashGenerator::updateDigestUseDependencies(const GlobalObject *GO,
-                                                      MD5 &GOHash,
-                                                      unsigned GODepth,
-                                                      const GOVec &Dependencies,
-                                                      bool UseRepoDefinitionMD)
+auto ModuleHashGenerator::updateDigestUseState(const GlobalObject *GO,
+                                               MD5 &GOHash, unsigned GODepth,
+                                               const GODigestState &State,
+                                               bool UseRepoDefinitionMD)
     -> std::tuple<size_t, DigestType> {
   auto LoopPoint = std::numeric_limits<size_t>::max();
-  for (const GlobalObject *const G : Dependencies) {
+  for (const GlobalObject *const G : llvm::concat<const GlobalObject *const>(
+           make_range(State.Contributions.begin(), State.Contributions.end()),
+           make_range(State.Dependencies.begin(), State.Dependencies.end()))) {
```

Comment:
This proposed code looks a great deal like the code that was removed by the work for PR #89. Could you provide some reassurance that we’re not going to re-introduce this same problem?

Reply:
> 
> 
> This proposed code looks a great deal like the code that was removed by the work for PR #89. 
>
Yes, the PR #89 tried to improve the performance by considering the `contribution` and `dependencies` separately, but it introduced this new issue 110.

>
> Could you provide some reassurance that we’re not going to re-introduce this same problem?
>
I am assuming the same problem refers to the performance issue. The fix will drop performance around 10% (opt&backend) for upstream llvm-project. 

I tried to keep `contributions` and dependencies` loops separately but it doesn't work since `contributions` and `dependencies` can form the loop and need to be consider at the same time.

Reply:
I am working on another possible fix and will update later.

Reply:
I have updated the patch with another method: update the `Contribution` and `Dependencies` hash separately.

The patch includes the two commits.

1) commit (https://github.com/SNSystems/llvm-project-prepo/pull/111/commits/6477be8eec7933769f8a2481fc8337a099e57433) fixed the incorrect hash calculation. It will increase the repo hash calculation time as expected.
2) commit (https://github.com/SNSystems/llvm-project-prepo/pull/111/commits/b41f6ea8668886fdc275ee176699840a7718306e) improved the repo hash calculation time.

The upstream llvm-project (commit: `a822ec75`) is used to measure the performance improvement.

The backend compilation times of the repo compilers are shown below.

master branch (commit: d93faf2453980a7234899c3f58303ab00e558d2b): `17166.0s`. 
issue110 branch (commit: 6477be8eec7933769f8a2481fc8337a099e57433): `18834.0s`. 
issue110 branch (commit: b41f6ea8668886fdc275ee176699840a7718306e): `17178.2s`.